File: app/src/main/java/wangdaye/com/geometricweather/ui/widget/DayNightShaderWrapper.java
package wangdaye.com.geometricweather.ui.widget;

import android.graphics.Shader;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.util.Arrays;

public class DayNightShaderWrapper {

    @Nullable private Shader shader;

    private int targetWidth;
    private int targetHeight;

    private boolean lightTheme;
    private int[] colors;

    public DayNightShaderWrapper(int targetWidth, int targetHeight) {
        this(null, targetWidth, targetHeight, true, new int[0]);
    }

    public DayNightShaderWrapper(@Nullable Shader shader, int targetWidth, int targetHeight,
                                 boolean lightTheme, @NonNull int[] colors) {
        setShader(shader, targetWidth, targetHeight, lightTheme, colors);
    }

    public boolean isDifferent(int targetWidth, int targetHeight,
                               boolean lightTheme, @NonNull int[] colors) {
        if (this.shader == null
                || this.targetWidth != targetWidth
                || this.targetHeight != targetHeight
                || this.lightTheme != lightTheme
                || this.colors.length != colors.length) {
            return true;
        }

        for (int i = 0; i < colors.length; i ++) {
            if (this.colors[i] != colors[i]) {
                return true;
            }
        }

        return false;
    }

    public void setShader(@Nullable Shader shader, int targetWidth, int targetHeight,
                          boolean lightTheme, @NonNull int[] colors) {
        this.shader = shader;
        this.targetWidth = targetWidth;
        this.targetHeight = targetHeight;
        this.lightTheme = lightTheme;
        this.colors = Arrays.copyOf(colors, colors.length);
    }

    @Nullable
    public Shader getShader() {
        return shader;
    }

    public int getTargetWidth() {
        return targetWidth;
    }

    public int getTargetHeight() {
        return targetHeight;
    }

    public boolean isLightTheme() {
        return lightTheme;
    }
}

File: app/src/main/java/org/breezyweather/common/ui/widgets/DayNightShaderWrapper.kt
/**
 * This file is part of Breezy Weather.
 *
 * Breezy Weather is free software: you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as published by the
 * Free Software Foundation, version 3 of the License.
 *
 * Breezy Weather is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public
 * License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with Breezy Weather. If not, see <https://www.gnu.org/licenses/>.
 */

package org.breezyweather.common.ui.widgets

import android.graphics.Shader

class DayNightShaderWrapper @JvmOverloads constructor(
    targetWidth: Int,
    targetHeight: Int,
    lightTheme: Boolean = true,
    colors: IntArray = IntArray(0),
    shader: Shader? = null,
) {
    var shader: Shader? = null
        private set
    var targetWidth = 0
        private set
    var targetHeight = 0
        private set
    var isLightTheme = false
        private set
    private var mColors: IntArray = IntArray(0)

    init {
        setShader(shader, targetWidth, targetHeight, lightTheme, colors)
    }

    fun isDifferent(
        targetWidth: Int,
        targetHeight: Int,
        lightTheme: Boolean,
        colors: IntArray,
    ): Boolean {
        if (shader == null ||
            this.targetWidth != targetWidth ||
            this.targetHeight != targetHeight ||
            isLightTheme != lightTheme ||
            mColors.size != colors.size
        ) {
            return true
        }
        for (i in colors.indices) {
            if (mColors[i] != colors[i]) {
                return true
            }
        }
        return false
    }

    fun setShader(
        shader: Shader?,
        targetWidth: Int,
        targetHeight: Int,
        lightTheme: Boolean,
        colors: IntArray,
    ) {
        this.shader = shader
        this.targetWidth = targetWidth
        this.targetHeight = targetHeight
        isLightTheme = lightTheme
        mColors = colors.copyOf(colors.size)
    }
}
